### PR TITLE
Del data bug

### PR DESCRIPTION
**Pull request by dstuebe**
Fixed the delete data bug and improve testing and stability of the metadata cache.

### CODE REVIEW

File: ion/core/object/gpb_wrapper.py [hunk 1]
```diff
@@ -1005,7 +1005,7 @@ class Wrapper(object):
 
         self.recurse_count.count += 1
         local_cnt = self.recurse_count.count
-        log.debug('Entering Recurse Commit: recurse counter - %d, Object Type - %s, child links - %d, objects to commit - %d:\n %s' %
+        log.debug('Entering Recurse Commit: recurse counter - %d, Object Type - %s, child links - %d, objects to commit - %d, Modified - %s' %
```
Comment (dstuebe) [inline ion/core/object/gpb_wrapper.py:1008]: This is not an AIS thing - but I found it here...

File: ion/integration/ais/app_integration_service.py [hunk 2]
```diff
@@ -94,11 +94,11 @@ class AppIntegrationService(ServiceProcess):
         yield self.metadataCache.loadDataSources()
 
         log.info('instantiating DatasetUpdateEventSubscriber')
-        self.dataset_subscriber = DatasetUpdateEventSubscriber(self, process = self)
```
Comment (dstuebe) [inline ion/integration/ais/app_integration_service.py:97]: This is important!

File: ion/integration/ais/common/metadata_cache.py [hunk 3]
```diff
@@ -78,7 +78,6 @@ VISIBILITY = 'visibility'
 
 class MetadataCache(object):
     
-    __metadata = {}
```
Comment (dstuebe) [inline ion/integration/ais/common/metadata_cache.py:81]: This should not be a class attribute - it causes problems in testing and when running more than one instance of the AIS service in a single CC.

File: ion/integration/ais/findDataResources/findDataResources.py [hunk 4]
```diff
@@ -65,11 +65,13 @@ if DNLD_FILE_TYPE is None:
     log.error('DNLD_FILE_TYPE not set in ion.config or ionlocal.config!  Using %s' %(DNLD_FILE_TYPE))
 
 class DatasetUpdateEventSubscriber(DatasetChangeEventSubscriber):
-    def __init__(self, ais, *args, **kwargs):
```
Comment (dstuebe) [inline ion/integration/ais/findDataResources/findDataResources.py:68]: This is important - the process is passed as a kwarg - the old method of instantiation can cause problems...

File: ion/integration/ais/findDataResources/findDataResources.py [hunk 5]
```diff
@@ -82,22 +84,10 @@ class DatasetUpdateEventSubscriber(DatasetChangeEventSubscriber):
...
         #
-        # Check the cache to see if there's currently metadata for this
-        # datasetID
-        #
-        dSetMetadata = yield self.metadataCache.getDSetMetadata(dSetResID)
-
-        #
-        # If dataset does not exist, this must be a new dataset; skip the
-        # delete step.
+        # Delete the dataset
```
Comment (dstuebe) [inline ion/integration/ais/findDataResources/findDataResources.py:87]: Changed it to always delete on update - simpler

File: ion/integration/ais/findDataResources/findDataResources.py [hunk 6]
```diff
@@ -107,13 +97,17 @@ class DatasetUpdateEventSubscriber(DatasetChangeEventSubscriber):
 
         log.debug("DatasetUpdateEventSubscriber event for dsetID: %s exit" %(dSetResID))
 
-                
+        if not self._hook_for_testing.called:
+            self._hook_for_testing.callback(True)
```
Comment (dstuebe) [inline ion/integration/ais/findDataResources/findDataResources.py:101]: A deferred hook for deterministic testing!

File: ion/integration/ais/findDataResources/findDataResources.py [hunk 7]
```diff
@@ -168,32 +142,40 @@ class DatasourceUpdateEventSubscriber(DatasourceChangeEventSubscriber):
...
-            if (tempDSourceResID is None) or (tempDSourceResID != dSourceResID):
-                log.error('DSOURCE_ID was none or changed for dSetID %s; refreshing dataset' %(dSetResID))
+                #
+                # Now  load the dataset metadata
+                #
+                log.debug('DatasourceUpdateEventSubscriber loading dataset %s metadata into cache' %(dSetResID))
+                yield self.metadataCache.putDSetMetadata(dSetResID)
+
+            elif dSet.get('DSOURCE_ID') != dSourceResID:
+                log.info('DataSet Metadata DSOURCE_ID was none or did not match for dSetID %s; refreshing dataset' %(dSetResID))
```
Comment (dstuebe) [inline ion/integration/ais/findDataResources/findDataResources.py:163]: Cleaner way to write this business logic.

File: ion/integration/ais/manage_data_resource/manage_data_resource.py [hunk 8]
```diff
@@ -304,23 +309,26 @@ class ManageDataResource(object):
...
-                deletions.append(data_set_resource_id)
 
             log.info("putting all resource changes in one big transaction, " \
                          + str(len(delete_resources)))
             yield self.rc.put_resource_transaction(delete_resources)
             log.info("Success!")
 
-            log.info("creating event to signal caching service")
-            yield self.pub_dsrc.create_and_publish_event(origin=datasrc_resource.ResourceIdentity,
+            for datasrc_resource in delete_resources:
```
Comment (dstuebe) [inline ion/integration/ais/manage_data_resource/manage_data_resource.py:327]: This is the delete multiple sources bug!

**Comment by deverett** [comment]
Reviewed.

**Pull request merged**